cmd/agent/agent.go

```diff
@@ -40,6 +41,10 @@ func NewAgentCmd() *cobra.Command {
 			return internal.InitConfig("agent")
 		},
 	}
+
+	startCmd.Flags().StringVar(&bindIP, "bind-ip", "", "Private address to which the trento-agent should be bound for internal communications. Reachable by the other hosts, including the trento server.")
+	startCmd.MarkFlagRequired("bind-ip")
```

Comment:
This is very confusing. 
I see this configuration item only used when invoking `collector.Client.Publish()`, which issues a request towards the server.
So, as far as I can see, we're not actually binding to any network interface and, when Consul will be out of the picture, we won't have inbound traffic anymore and we won't need the agent host to be reachable at this IP address anymore. What gives?

Maybe we need to rethink the need to ask the user the ip address, because the server will inherently know this IP address from the request being issued by the agent collector client.

Reply:
> I see this configuration item only used when invoking collector.Client.Publish(), which issues a request towards the server.
>
mh :thinking: I am confused :sweat_smile: that is used in `./cmd/agent/config.go:52`
```
return &agent.Config{
    ...
    BindIP:          viper.GetString("bind-ip"),
    ...
}, nil
```
However.

> So, as far as I can see, we're not actually binding to any network interface 
>
Correct.

> and, when Consul will be out of the picture, we won't have inbound traffic anymore and we won't need the agent host to be reachable at this IP address anymore.
>
Partially correct I think, the runner needs to ssh to the hosts and it can be either via an IP or a resolved hostname.
When dealing with the hosts we usually have multiple IPs, yet I didn't find a way get the correct one for Ansible usage.
When installing via the `./install-agent.sh` we require a `--agent-bind-ip` which was used for consul, now we keep it and we use it for another purpose. 
Users installing the rpm have to put a `bind-ip: 0.0.0.0` in their `agent.yaml`, and this will be documented.

Do you have any other options? How can we deal with this differently, now? Maybe I am missing something :sweat_smile:

Reply:
> 
> mh thinking I am confused now sweat_smile that is used in [`./cmd/agent/config.go:52`](https://github.com/trento-project/trento/pull/546/files#diff-a22afc9279c2983d19ef8def93e25c50e43490a142d26b765296152a96c35d39R53)
> 
> ```
> return &agent.Config{
>     ...
>     BindIP:          viper.GetString("bind-ip"),
>     ...
> }, nil
> ```

Exactly, and that `BindIP` field in turn is never used to actually start any listening service bound to any network address. It's just sent as a field of a JSON to the server.

I wasn't clear in keeping SSH out of this picture: the SSH server is not bound to any particular IP address by the Agent process. How SSH is configured is out of the scope of our intervention; besides, the SSH server is usually not bound to any particular network address (i.e. 0.0.0.0:22), unlike Consul, which specifically needs a single interface.

My point is that the Agent process doesn't have to bind anything, because it's not a service, it's a client. SSH being there is a pre-existing condition we have no control over.

Yes, we can ask the user to specify which IP address to "advertise", but that's not a "bind" address, because it's actually not "bound" to anything by the Agent process.

Reply:
> Exactly, and that BindIP field in turn is never used to actually start any listening service bound to any network address. It's just sent as a field of a JSON payload to the server.
>
Okay, now its clearer what you mean.
That IP address is actually not used to _bind_ anything, it is published as part of the `HostDiscovery`.
We can change the name to something different.

> I wasn't clear in keeping SSH out of this picture: the SSH server is not bound to any particular IP address by the Agent process. How SSH is configured is out of the scope of our intervention...
>
> My point is that the Agent process doesn't have to bind anything, because it's not a service, it's a client. SSH being there is a pre-existing condition we have no control over.
>
I see, yet I need to answer the question: where should I ssh when running Ansible?

A little backgournd on the runner:
We provide an `Inventory` to Ansible for check execution, that is basically a list of hosts where to connect to and which checks to execute - nodes can also be grouped by cluster.

Here's a `Node`
```
type Node struct {
	Name        string // <-- hostname eg: vmhana01, vmnetweaver02
	AnsibleHost string <-- An IP, which? It comes to be the `BindIP` (better name to be found here)
	AnsibleUser string <-- ssh connection user
	Variables   map[string]interface{} <-- list of the checks to be executed (and maybe more, but not sure)
}
```
These Nodes get populated with the result obtained by calling `GET /api/clusters/settings` and are part of the `Inventory` used by ansible.

I don't know Ansible and I am not aware of the rules (is a `Name` always expected? What about the `AnsibleHost`? can it be empty? AnsibleUser too? Will ansible ssh to the hostname or the IP?) - and I need to do the safest thing, being requiring explicitly that IP to the user, as we already do (simply it will change semantic)

So what we can do is:
- reword `bind-ip` flag to something different (`ssh-host`, `connection-ip`, `runner-connection-hos`, whatever) pls suggest
- document the newly added flag

What do you think?

Reply:
I think we can go on with the following: `bind-ip` can become `ssh-address`. 

I guess this will become the content of the `AnsibleHost` field.
I don't think we'll need the `Name` field in that struct, because I think that's currently there only because Consul's catalog reasons in terms of host names; Ansible doesn't: it needs just an address, which is not necessarily always just an IP, it could be a FQDN and it could have a `:port` suffix.

Reply:
Okay.

I made the change in [this commit](https://github.com/trento-project/trento/pull/546/commits/de55e1b65c6bcd15b18646210b4834094e1c3591) and it touches quite a lot of things.
Take a look and share your feelings.
@fabriziosestito @dottorblaster I'd like to know what you think.

We're completely changing the semantic from a more generic `bind-ip` to a really specific `ssh_address`.
If we're okay with this I'll just go ahead (refining a bit more the documentation)

Still, I feel that this change is too drastic from a semantic perspective, especially at this time. I am afraid it would add just more cognitive complexity, not sure what you think.

So here's the current definition we have for `--agent-bind-ip`
> the private address to which the trento-agent should be bound for internal communications.
> This is an IP address that should be reachable by the other hosts, including the trento server.
> 

To find a good naming we should start from that definition: what is that IP address we expect representing?

A variant based on what we have could be:
> the address to which the trento-agent should be reachable by the runner for check execution.
> 
Please feel free to contribute.

We could leave out the details of _how_ (ssh, http, whatever)
Or, if that `ssh` detail is really the key, then we go for it.

So, as a tradeoff, I see possible to change the meaning and the name to `check-execution-address`, `check-address`, `runner-check-address` or something similar, please help with naming :sweat_smile: 

Maybe `ssh_address` is just good enough, I am still not fully convinced, though.

Reply:
@nelsonkopliku  imho `ssh-address` is good enough and enough explicit for now

Reply:
Ok for me keeping `ss-address` for now.